index.js

```diff
@@ -793,25 +797,28 @@ Command.prototype.normalize = function(args) {
 Command.prototype.parseArgs = function(args, unknown) {
   var name;
 
+  if (this.listeners('program-action').length && this.listeners('command:*').length) {
+    console.error("Can't have listeners for both command:* and for the main program");
+    this._exit(1);
+  }
+
   if (args.length) {
     name = args[0];
     if (this.listeners('command:' + name).length) {
       this.emit('command:' + args.shift(), args, unknown);
-    } else {
+    } else if (this.listeners('command:*').length) {
       this.emit('command:*', args, unknown);
+    } else {
+      this.emit('program-action', args, unknown);
     }
   } else {
     outputHelpIfNecessary(this, unknown);
-
     // If there were no args and we have unknown options,
     // then they are extraneous and we need to error.
     if (unknown.length > 0 && !this.defaultExecutable) {
       this.unknownOption(unknown[0]);
     }
-    if (this.commands.length === 0 &&
-        this._args.filter(function(a) { return a.required; }).length === 0) {
-      this.emit('command:*');
-    }
+    this.emit('program-action');
```

Comment:
The filter was to avoid calling program action handler when it has a required option which is missing, like:
`.arguments(<file>)`. So I think that is still needed.

I am not sure about the `this.commands.length` check, still working out what that was covering and why.

Reply:
Oh, we have an explicit test that program action handler not called if there is a subcommand and no arguments are specified. The test name is missing part of the scenario: "when no arguments then asterisk action not called'.

I want to read back through the issues to see if that was a key behaviour, or just the safest way of adding the call to program handler in special circumstances.

Reply:
Ah ha! The `this.commands.length` was to avoid calling `.command('*')` handler when no parameter from https://github.com/tj/commander.js/pull/844/files

I do not think we need that check now, I do not see a reason change emit of `program-action` bases on whether there are commands.

Reply:
Note: I added the filter back in one of my Pull Requests.